pslogin/src/main/scala/WorldSessionActor.scala

```diff
@@ -9239,39 +9118,24 @@ class WorldSessionActor extends Actor
       owner.Inventory.Items.foreach { case InventoryItem(obj, index) =>
         clobber.append(EncodeLoadoutCLOBFragment(obj, index))
       }
-      clobber.mkString
-    }
-    //database
-    Database.getConnection.connect.onComplete {
-      case scala.util.Success(connection) =>
-        Database.query(connection.sendPreparedStatement(
-          "SELECT id, exosuit_id, name, items FROM loadouts where characters_id = ? AND loadout_number = ?", Array(charId, line).toIndexedSeq
-        )).onComplete {
-          case scala.util.Success(queryResult) =>
-            queryResult match {
-              case row: ArrayRowData => // Update
-                connection.sendPreparedStatement(
-                  "UPDATE loadouts SET exosuit_id=?, name=?, items=? where id=?", List(exosuitId, label, clob.drop(1), row(0))
-                ).onComplete {
-                  case _ =>
-                    if(connection.isConnected) connection.disconnect
-                }
-                Thread.sleep(50)
-              case _ => // Save
-                connection.sendPreparedStatement(
-                  "INSERT INTO loadouts (characters_id, loadout_number, exosuit_id, name, items) VALUES(?,?,?,?,?) RETURNING id",
-                  List(charId, line, exosuitId, label, clob.drop(1))
-                ).onComplete {
-                  case _ =>
-                    if(connection.isConnected) connection.disconnect
-                }
-                Thread.sleep(50)
-            }
-          case scala.util.Failure(e) =>
-            if(connection.isConnected) connection.disconnect
-            log.error(s"SaveLoadoutToDB: query failed - ${e.getMessage}")
-        }
-      case scala.util.Failure(e) =>
+      clobber.mkString.drop(1)
+    }
+
+    val result = for {
+      loadouts <- ctx.run(query[persistence.Loadout].filter(_.charactersId == lift(owner.CharId)).filter(_.loadoutNumber == lift(line)))
+      loadout <- loadouts.headOption match {
+        case Some(loadout) =>
+          ctx.run(query[persistence.Loadout].filter(_.id == lift(loadout.id)).update(_.exosuitId -> lift(owner.ExoSuit.id), _.name -> lift(label), _.items -> lift(items)))
+        case None =>
+          // FIXME id is long in objects but int in schema
+          val charId: Int = owner.CharId.toInt
```

Comment:
A lot of IDs in the game objects are longs. Shouldn't really be necessary, do we make them int or do we make the db values longs?

Reply:
Character ID is always represented as 32-bits in the packets and you can't extract that into an `Int`.

Reply:
But `Int` is 32 bit..?

Reply:
`Scala` `Int` is signed.  `scodec` treats a stream of bits as unsigned.  It has to take a 32-bit number and push it into the datatype with the next highest bit length to preserve the value.
The `uint32` transcoder resolves into a `Long` value.

Reply:
Ah TIL there's no `UInt` in scala. Seems like postgres doesn't have it either, I'll make them longs in the schema when I'm done with migrations.

Reply:
If you wanted to write a 32-bit -> `Int` transcoder, you'd use `bool :: uint(31)` and manually factor for the sign and two's complement with an `xmap`.

Reply:
Right, could also do that, although I think a custom encoder would require a custom type as well. I'll look into it.

Reply:
Er, actually, I forgot.  There are transcoders for signed numbers, I just never use them. The native `intL32` goes into an `Int`.
Most shorter fields utilize their full bit length and numbers are almost never negative.  In terms of character id, the only difference between using `Int` and `Long` is the absurd number of unique characters that can be created.